sw/device/lib/dif/dif_hmac.c

```diff
@@ -99,110 +120,148 @@ dif_hmac_result_t dif_hmac_init(const dif_hmac_config_t *config,
-      mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
-                                        HMAC_CFG_ENDIAN_SWAP);
-      break;
-    default:
-      return kDifHmacError;
+  // Restore interrupt state.
+  mmio_region_write32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET, state);
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_irq_control(const dif_hmac_t *hmac,
+                                       dif_hmac_interrupt_t irq_type,
+                                       dif_hmac_enable_t enable) {
+  if (hmac == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  ptrdiff_t offset;
+  if (!irq_bit_offset_get(irq_type, &offset)) {
+    return kDifHmacError;
+  }
+
+  // Enable/Disable interrupt state.
+  if (enable == kDifHmacEnable) {
+    mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                                    offset);
+  } else {
+    mmio_region_nonatomic_clear_bit32(hmac->base_addr,
+                                      HMAC_INTR_STATE_REG_OFFSET, offset);
   }
 
   return kDifHmacOk;
 }
 
-dif_hmac_result_t dif_hmac_digest_endianness_set(
-    const dif_hmac_t *hmac, dif_hmac_endianness_t endianness) {
+dif_hmac_result_t dif_hmac_irq_force(const dif_hmac_t *hmac,
+                                     dif_hmac_interrupt_t irq_type) {
   if (hmac == NULL) {
     return kDifHmacBadArg;
   }
 
-  switch (endianness) {
-    case kDifHmacEndiannessBig:
-      mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
-                                      HMAC_CFG_DIGEST_SWAP);
-      break;
-    case kDifHmacEndiannessLittle:
-      mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
-                                        HMAC_CFG_DIGEST_SWAP);
-      break;
-    default:
-      return kDifHmacError;
+  ptrdiff_t offset;
+  if (!irq_bit_offset_get(irq_type, &offset)) {
+    return kDifHmacError;
   }
 
+  // Force the requested interrupt.
+  mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_INTR_TEST_REG_OFFSET,
+                                  offset);
+
   return kDifHmacOk;
 }
 
-dif_hmac_result_t dif_hmac_key_set(const dif_hmac_t *hmac, const uint8_t *key) {
+dif_hmac_result_t dif_hmac_mode_hmac_start(const dif_hmac_t *hmac,
+                                           const uint8_t *key) {
   if (hmac == NULL || key == NULL) {
     return kDifHmacBadArg;
   }
 
   load_key(hmac, key);
 
+  // Set HMAC to process in HMAC mode (not SHA256-only mode).
+  mmio_region_nonatomic_set_mask32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                   HMAC_CFG_SHA_EN | HMAC_CFG_HMAC_EN, 0);
+
+  // Begin HMAC operation.
+  mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_START);
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_mode_sha256_start(const dif_hmac_t *hmac) {
+  if (hmac == NULL) {
+    return kDifHmacBadArg;
+  }
+
```

Comment:
Turns out there are a couple of issues with these new start functions, I believe I've addressed everything now.

Reply:
I think it would be better to have a `dif_hmac_finish` which combines `dif_hmac_digest_read` and clearing `CFG.SHA_EN` and `CFG.HMAC_EN`. Then it becomes clearer when a transaction is in progress (between the start and finish calls). It would mean you only had one opportunity to read the digest, but I don't think that's a problem, if you need it for longer read it into a buffer and pass that around.

Let's file this in the follow-up issue, not to block this PR.